### PR TITLE
Allow OMPI to compile on OSX with gcc13

### LINKED ISSUE #12108: Build error on macOS (`PATH_MAX` undeclared)
I am building OpenMPI 5.0.0 on macOS with a MacPorts-installed GCC 13.2.

My configure options are
```
$ ./configure --prefix=$HOME/openmpi-5.0.0 --with-libevent=internal CC=gcc CXX=g++ FC=gfortran
```

This fails with the error message
```
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include/stdio.h:409,
                 from ./include/mpl_base.h:14,
                 from ./include/mpl.h:9,
                 from src/str/mpl_str.c:6:
src/str/mpl_str.c: In function 'MPL_create_pathname':
src/str/mpl_str.c:413:37: error: 'PATH_MAX' undeclared (first use in this function); did you mean 'AF_MAX'?
  413 |         MPL_snprintf(dest_filename, PATH_MAX, "%s/%s.%u.%u%c", dirname, prefix,
      |                                     ^~~~~~~~
```

This refers to the file `3rd-party/romio341/mpl/src/str/mpl_str.c` (I wish that was reported directly in the error message). Adding the line `#include <sys/syslimits.h>` to the beginning of this file resolves the error.

### LINKED ISSUE #12109: Build error on macOS (`SSIZE_MAX` undefined)
I am building OpenMPI 5.0.0 on macOS with a MacPorts-installed GCC 13.2.

My configure options are
```
$ ./configure --prefix=$HOME/openmpi-5.0.0 --with-libevent=internal CC=gcc CXX=g++ FC=gfortran
```

This fails with the error message
```
get_elements_x.c: In function 'PMPI_Get_elements_x':
get_elements_x.c:77:63: error: 'SSIZE_MAX' undeclared (first use in this function); did you mean 'RSIZE_MAX'?
   77 |         if (OMPI_SUCCESS == ret && internal_count <= (size_t) SSIZE_MAX) {
      |                                                               ^~~~~~~~~
      |                                                               RSIZE_MAX
get_elements_x.c:77:63: note: each undeclared identifier is reported only once for each function it appears in
```

Adding the lines
```
#ifndef SSIZE_MAX
#define SSIZE_MAX LONG_MAX
#endif
```
after the `#include <limits.h>` resolves the issue for me.

I'm sure there is a more elegant way, probably requiring a certain POSIX level, but I could not figure it out.

### LINKED ISSUE #12110: Build error on macOS: `IOV_MAX` undefined
I am building OpenMPI 5.0.0 on macOS with a MacPorts-installed GCC 13.2.

My configure options are

$ ./configure --prefix=$HOME/openmpi-5.0.0 --with-libevent=internal CC=gcc CXX=g++ FC=gfortran
This fails with the error message
```
fbtl_posix_preadv.c: In function 'mca_fbtl_posix_preadv_generic':
fbtl_posix_pwritev.c: In function 'mca_fbtl_posix_pwritev_generic':
fbtl_posix_preadv.c:282:30: error: 'IOV_MAX' undeclared (first use in this function); did you mean 'INT_MAX'?
  282 |                 (iov_count < IOV_MAX ) ){
      |                              ^~~~~~~
      |                              INT_MAX
fbtl_posix_pwritev.c:286:31: error: 'IOV_MAX' undeclared (first use in this function); did you mean 'INT_MAX'?
  286 |                  (iov_count < IOV_MAX )) {
      |                               ^~~~~~~
      |                               INT_MAX
fbtl_posix_preadv.c:282:30: note: each undeclared identifier is reported only once for each function it appears in
fbtl_posix_pwritev.c:286:31: note: each undeclared identifier is reported only once for each function it appears in
```

### PR DESCRIPTION
**Pull request by bosilca**
Thanks @eschnett  for reporting the 3 issues related to gcc13.

I am not entirely convinced the root cause is on OMPI side. It seems to me that some versions of gcc13.2 fail to define critical POSIX constants/defines (IOV_MAX, PATH_MAX and SSIZE_MAX) as well as their minimal values. Digging a little around it seems gcc13.2 (at least the MacPort variant) fails to use the OSX system headers to get these constants, aka. basically missing an #include_next on limits.h (this #include_next existed in gcc12).

Anyway, until we figure out the real solution I need to be able to compile on my laptop with gcc13.

Fixes #12108 
Fixes #12109
Fixes #12110

### CODE REVIEW

File: opal/include/opal_config_bottom.h
Comment (jsquyres) [inline opal/include/opal_config_bottom.h:573]: Just to calrify: is this a gcc 13 issue, a gcc 13.2 issue, a macports 13.2.0 issue, or ...?

Also, is `SSIZE_MAX` guaranteed to be a macro?  E.g., could it be an actual symbol such that `#ifndef SSIZE_MAX` would be false, but using `SSIZE_MAX` would still be fine?

**Review by jsquyres — COMMENTED** [review]
Can you cite @eschnett somewhere?  He did report these issues.

Per https://github.com/open-mpi/ompi/issues/12110#issuecomment-1824464914, I don't actually have problems compiling OMPI with Homebrew-installed gcc 13.2.0.  🤷‍♂️ 

But most of these changes look good, anyway -- the only quibble I have is with `SSIZE_MAX` (noted above).

File: opal/include/opal_config_bottom.h
Comment (bosilca) [reply opal/include/opal_config_bottom.h:573]: According to POSIX, SSIZE_MAX is supposed to be a constant expression suitable for use in #if. In practice we found an instance where it does not exists, so I imagine it could be possible that it might not be a constant expression.

Comment (ggouaillardet) [reply opal/include/opal_config_bottom.h:573]: This one is a bit tricky...
It works on my Mac with Command Line Tools 13.2 and GCC 13.2.0 provided by brew
But it does not work on an other Mac with Command Line Tools 14.0 and GCC 13.2.0 provided by brew

At first glance, this is because Command Line Tools 14.0 does not provide `limits.h` any more.
That being said, from `/usr/local/Cellar/gcc/13.2.0/lib/gcc/current/gcc/x86_64-apple-darwin22/13/include/limits.h`
```
#ifdef _GCC_NEXT_LIMITS_H
#include_next <limits.h>                /* recurse down to the real one */
#endif
```

I would expect `_GCC_NEXT_LIMITS_H` is **not** defined when recent Command Line Tools are used.
But if I manually `#undef _GCC_NEXT_LIMITS_H`, the `SSIZE_MAX` macro is still undefined.

I've heard several folks from `brew` complaining Apple basically does not care about these issues, so we might want to be pragmatic here and merge this PR.

**Review by jsquyres — APPROVED** [review]

File: opal/include/opal_config_bottom.h
Comment (jsquyres) [inline opal/include/opal_config_bottom.h:573]: Actually, could you fix this comment so that we don't have this stored in history?  It's not really a gcc issue; it's a platform issue (e.g., homebrew gcc 13 doesn't seem to have this problem).

File: ompi/mca/fs/base/fs_base_get_parent_dir.c
Comment (jsquyres) [inline ompi/mca/fs/base/fs_base_get_parent_dir.c]: Looks like an errant copyright update for a file with no changes.

So we don't hold this PR up further, I'll push an update removing this line.

**Review by jsquyres — APPROVED** [review]

**Comment by bosilca** [comment]
all the commits are unverified now. let me try to fix this.

**Comment by jsquyres** [comment]
> all the commits are unverified now. let me try to fix this.

Sorry, that was probably just because I pushed last night (i.e., I became the committer on those commits).

The NVIDIA CI was failing last night; let's see if it passes today...

**Pull request merged**
